### ISSUE TITLE
CountryType doesn't display correctly if parent has the same class name

### ISSUE THREAD

**Issue by foaly-nr1** [symfony/symfony#18953]
TL;DR: I added code to reproduce this issue in the branch `bug-country-type` of [my Symfony Standard Edition fork](https://github.com/foaly-nr1/symfony-standard/tree/bug-country-type).

When Symfony's `CountryType` field is added to a form class `CountryType` it displays as a input type text instead of a select field. In the following two examples, the only difference is the form's class name. When displayed below each other they look like this:

![screen shot 2016-06-02 at 14 45 08](https://cloud.githubusercontent.com/assets/282458/15747177/682979e0-28d1-11e6-9bd9-71d82700f7a5.png)

### Not working

```php
namespace AppBundle\Form\Type;

use Symfony\Component\Form\AbstractType;
use Symfony\Component\Form\FormBuilderInterface;
use Symfony\Component\Form\Extension\Core\Type;
use Symfony\Component\OptionsResolver\OptionsResolver;

class CountryType extends AbstractType
{
    /**
     * {@inheritdoc}
     */
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder
            ->add('countryCode', Type\CountryType::class)
            ->add('capacity', Type\NumberType::class)
            ->add('save', Type\SubmitType::class)
        ;
    }

    /**
     * {@inheritdoc}
     */
    public function configureOptions(OptionsResolver $resolver)
    {
        $resolver->setDefaults([
            'data_class' => 'AppBundle\Entity\Country',
        ]);
    }
}
```

### Working

```php
namespace AppBundle\Form\Type;

use Symfony\Component\Form\AbstractType;
use Symfony\Component\Form\FormBuilderInterface;
use Symfony\Component\Form\Extension\Core\Type;
use Symfony\Component\OptionsResolver\OptionsResolver;

class CountryWorkingType extends AbstractType
{
    /**
     * {@inheritdoc}
     */
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder
            ->add('countryCode', Type\CountryType::class)
            ->add('capacity', Type\NumberType::class)
            ->add('save', Type\SubmitType::class)
        ;
    }

    /**
     * {@inheritdoc}
     */
    public function configureOptions(OptionsResolver $resolver)
    {
        $resolver->setDefaults([
            'data_class' => 'AppBundle\Entity\Country',
        ]);
    }
}
```

**Comment by stof** [comment]
The other working solution is to overwrite ``getBlockPrefix`` to avoid relying on the short class name (as this will generate ``country`` for your own type, which should indeed not have the same rendering than the ``country`` rendering shipped in Symfony)

**Comment by javiereguiluz** [comment]
@stof can this be considered an edge-case bug of the Form component or is it an intended feature? Thanks!

**Comment by foaly-nr1** [comment]
@stof @javiereguiluz Maybe it'd good to add a section to the docs pointing out that if a custom form type's (auto generated) block prefix matches that of any of the form types shipped with Symfony, it will not be rendered as expected – or in fact any any form type shipped with any registered bundle. I don't think it's too obvious.

I'm happy to create a pull request.

**Comment by stof** [comment]
@javiereguiluz reusing the block prefixes causes issues (a subset of the issues caused by reusing the same type name in 2.7).
The fact that AbstractType provides a default implementation of ``getBlockPrefix`` based on a non-unique value (the short class name rather than the FQCN) looks very unfortunate to me (as it makes issues even harder to spot than when users were defining ``getName`` explicitly).

**Comment by stof** [comment]
unfortunately, changing it now would also be a BC break :(

**Comment by javiereguiluz** [comment]
@foaly-nr1 thanks! The only reference to `getBlockPrefix()` that I've found in the docs is here: https://symfony.com/doc/current/cookbook/form/create_custom_field_type.html#creating-a-template-for-the-field

**Comment by javiereguiluz** [comment]
Closing this issue then as "fixed" because this is a feature that we can't change for BC reasons and we'll add a note in the docs.
